lib/writeGltf.js

```diff
@@ -38,40 +28,93 @@ function writeGltf(gltf, outputPath, isEmbedded, createDirectory, callback) {
                     throw err;
                 }
             }
+            else{
+                fs.writeFile(outputPath, JSON.stringify(gltf, undefined, 2), function (err) {
+                    if (err) {
+                        throw err;
+                    }
+                    if (callback) {
+                        callback();
+                    }
+                });     
+            }   
+        }
+    );
+}
 
-            fs.writeFile(outputPath, JSON.stringify(gltf, undefined, 2), function (err) {
-                if (err) {
-                    throw err;
+function writeSource(gltf, filePath, name, isEmbedded, callback) {
+    var objects = gltf[name];
+
+    //Iterate through each object and write out its uri
+    if (defined(objects)) {
+        var ids = Object.keys(objects);
+        async.each(ids, function(id, asyncCallback) {
+            var object = objects[id];
+            if (defined(object.extras) && defined(object.extras,source)) {
+                var source = object.extras.source;
+                delete object.extras.source;
+
+                //Write the source object as a data or file uri depending on the isEmbedded flag
+                if (isEmbedded) {
+                    var sourceDataUri = new dataUri();
+                    switch (name) {
+                        case 'images':
+                            sourceDataUri.format('.png', source);
+                            break;
+                        case 'shaders':
+                            sourceDataUri.format('.txt', source);
+                            break;
+                        default:
+                            throw new Error('Trying to embed invalid file type.');
+                    }
+                    object.uri = sourceDataUri.content;
+                    process.nextTick(function() {
+                        asyncCallback();
+                    });
                 }
+                else {
+                    //Choose file extension depending on object type
+                    var fileName = id;
+                    switch (name) {
+                        case 'images':
+                            fileName += '.png';
+                            break;
+                        case 'shaders':
+                            fileName += '.glsl';
+                            break;
+                        default:
+                            throw new Error('Trying to write out invalid file type.');
+                    }
+
+                    object.uri = fileName;
+                    var outputPath = path.join(filePath, fileName);
+                    fs.writeFile(outputPath, source, function (err) {
+                        if (err) {

```

Comment:
This `if...else` block can just be `asyncCallback(err)`.  If `err` is `undefined`, it is basically the same as `asyncCallback()`.